### PR TITLE
feat(schema): add pyarrow Extension "FileArray" & "FileType"

### PR DESCRIPTION
**Pull request by AChenQ**

### CODE REVIEW

**Comment by GravitiRob** [comment]
Task linked: [CU-2hd6ga0 使用 PyArrow 实现 File 类型的读取](https://app.clickup.com/t/2hd6ga0)

File: graviti/schema/extractors.py [hunk 1]
```diff
@@ -9,36 +9,46 @@
...
+from graviti.schema.pyarrow import FileType
 from graviti.utility.typing import NestedDict
 
 _T = TypeVar("_T")
 _D = Dict[str, Tuple[Iterator[_T], str]]
-_Extractor = Callable[[Dict[str, Any]], Iterator[_T]]
+_Extractor = Tuple[Callable[[Dict[str, Any]], Iterator[Any]], _T]
 Extractors = NestedDict[str, _Extractor[Any]]
 
+_PORTEX_TYPE_TO_NUMPY_DTYPE = {
```
Comment (Lee-000) [inline graviti/schema/extractors.py:22]: Not numpy dtype.

File: graviti/schema/pyarrow.py [hunk 2]
```diff
@@ -93,3 +95,30 @@ def __init__(
 
     def _to_pyobj(self) -> Dict[str, Any]:
         return {**super()._to_pyobj(), "url": self._url, "revision": self._revision}
+
+
+class FileArray(  # pylint: disable=too-few-public-methods
```
Comment (Lee-000) [inline graviti/schema/pyarrow.py:100]: Is the disable necessary?

**Review by Lee-000 — CHANGES_REQUESTED** [review]

File: graviti/schema/pyarrow.py [hunk 3]
```diff
@@ -93,3 +95,30 @@ def __init__(
...
+
+    def __getitem__(self, index: Union[int, slice]) -> Union[File, pa.Array]:
+        if isinstance(index, slice):
+            return super().__getitem__(index)
+
+        item = super().__getitem__(index).as_py()
+        return File(item["url"], item["checksum"])
+
+
+class FileType(BuiltinExtension):  # pylint: disable=too-few-public-methods
```
Comment (linjiX) [inline graviti/schema/pyarrow.py:113]: `FileType` is `external` or `builtin`?

File: graviti/schema/pyarrow.py [hunk 4]
```diff
@@ -93,3 +95,30 @@ def __init__(
...
+    """This class defines the PyArrow representation of FileType type."""
+
+    def __init__(self) -> None:
+        self.pa_type = pa.struct({"url": pa.string(), "checksum": pa.string()})
+        super().__init__("graviti.file", self.pa_type)
```
Comment (linjiX) [inline graviti/schema/pyarrow.py:118]: can we access `pa_type` by `self.storage_type`?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/schema/pyarrow.py [hunk 5]
```diff
@@ -93,3 +95,30 @@ def __init__(
...
+
+
+class FileArray(  # pylint: disable=too-few-public-methods
+    pa.ExtensionArray  # type: ignore[misc]
+):
+    """This class defines the PyArrow representation of FileArray."""
+
+    def __getitem__(self, index: Union[int, slice]) -> Union[File, pa.Array]:
+        if isinstance(index, slice):
+            return super().__getitem__(index)
```
Comment (linjiX) [inline graviti/schema/pyarrow.py:107]: if input is a slice? what will be returned? `pa.Array` or `pa.ExtensionArray` or `pa.FileArray`?

Comment (AChenQ) [reply graviti/schema/pyarrow.py:107]: `pa.FileArray`

File: graviti/schema/pyarrow.py (hunk 3 above)
Comment (AChenQ) [reply graviti/schema/pyarrow.py]: The `ExternalExtension` has the neessary parameter `url`, which means refer to the other repo?

File: graviti/schema/pyarrow.py (hunk 5 above)
Comment (linjiX) [reply graviti/schema/pyarrow.py:105]: should we update the return type to `pa.FileArray`?

Comment (linjiX) [reply graviti/schema/pyarrow.py:105]: Also should we add `overload` for `__getitem__`

File: graviti/schema/pyarrow.py [hunk 6]
```diff
@@ -93,3 +95,27 @@ def __init__(
...
+        return File(item["url"], item["checksum"])
+
+
+class FileType(BuiltinExtension):
+    """This class defines the PyArrow representation of FileType type."""
+
+    def __init__(self) -> None:
+        super().__init__("graviti.file", pa.struct({"url": pa.string(), "checksum": pa.string()}))
+
+    def __reduce__(self) -> Tuple[Any, Tuple[()]]:
```
Comment (Lee-000) [inline graviti/schema/pyarrow.py:117]: Why `__redunce__` instead of `__arrow_ext_serialize__`?

File: graviti/utility/lazy.py [hunk 7]
```diff
@@ -103,7 +103,11 @@ def update(self, pos: int, data: Any) -> None:
             data: The source data which needs to be input to the extractor.
 
         """
-        self.pages[pos] = pa.array(list(self._extractor(data)))
+        if isinstance(self._dtype, pa.PyExtensionType):
```
Comment (Lee-000) [inline graviti/utility/lazy.py:106]: `PyExtensionType` or `ExtensionType`?

File: graviti/schema/extractors.py [hunk 8]
```diff
@@ -86,7 +98,7 @@ def extractor(data: Dict[str, Any]) -> Iterator[Dict[str, Any]]:
             box2d.update(_extract_category_and_attribute(schema["items"], labels))
             yield box2d
 
-    return extractor
+    return extractor, pa.string()
```
Comment (Lee-000) [inline graviti/schema/extractors.py:101]: Would `pa.string()` cover the `box2ds` semantics?

File: graviti/schema/pyarrow.py [hunk 9]
```diff
@@ -93,3 +95,32 @@ def __init__(
...
+
+    def __getitem__(self, index: Union[int, slice]) -> Union[File, "FileArray"]:
+        if isinstance(index, slice):
+            return super().__getitem__(index)  # type: ignore[no-any-return]
+
+        item = super().__getitem__(index).as_py()
+        return File(item["url"], item["checksum"])
+
+
+class FileType(BuiltinExtension):
```
Comment (Lee-000) [inline graviti/schema/pyarrow.py:119]: Why not extend `ExtensionBase`?

File: graviti/utility/lazy.py [hunk 10]
```diff
@@ -103,7 +103,11 @@ def update(self, pos: int, data: Any) -> None:
             data: The source data which needs to be input to the extractor.
 
         """
-        self.pages[pos] = pa.array(list(self._extractor(data)))
+        if isinstance(self._dtype, pa.ExtensionType):
```
Comment (Lee-000) [inline graviti/utility/lazy.py:106]: Which one is better, `pa.ExtensionType` or `ExtensionBase`?

File: graviti/schema/pyarrow.py [hunk 11]
```diff
@@ -93,3 +95,38 @@ def __init__(
 
     def _to_pyobj(self) -> Dict[str, Any]:
         return {**super()._to_pyobj(), "url": self._url, "revision": self._revision}
+
+
+class GravitiExtension(ExtensionBase):  # pylint: disable=too-few-public-methods
```
Comment (Lee-000) [inline graviti/schema/pyarrow.py:100]: No need to disable the check.

**Review by Lee-000 — CHANGES_REQUESTED** [review]

**Review by Lee-000 — APPROVED** [review]

**Review by Lee-000 — APPROVED** [review]

**Review by linjiX — APPROVED** [review]

**Pull request merged**